src/pv.rs

```diff
@@ -61,6 +62,62 @@ impl PvPacket {
             private: std::ptr::null_mut(),
         }
     }
+
+    // check whether the new header can be attached to payload from the start of packet
+    fn is_addable(header_length: usize, packet: &PvPacket) -> bool {
+        let total_len: u32 = packet.size - packet.start;
+        let payload_len: u32 = packet.end - packet.start;
+
+        (header_length as u32) + payload_len <= total_len
+    }
+
+    // add header to payload
+    pub fn add_header(&mut self, header: Vec<u8>) -> Result<(), ()> {
+        let is_addable: bool = Self::is_addable(header.len(), self);
+
+        if is_addable {
+            unsafe {
+                copy(
+                    self.buffer.offset((self.start) as isize),
+                    self.buffer
+                        .offset((self.start + header.len() as u32) as isize),
+                    header.len(),
+                );
+                copy_nonoverlapping(
+                    header.as_ptr(),
+                    self.buffer.offset(self.start as isize),
+                    header.len(),
+                );
+            }
+            Ok(())
+        } else {
+            Err(())
+        }
+    }
+
+    // set all data to 0 and *end to *start
+    pub fn delete_data(&mut self) {
+        for i in self.start..self.end {
+            unsafe {
+                write(self.buffer.offset(i as isize), 0);
+            }
+        }
+
+        self.end = self.start;
+    }
+
+    // replace data with new data from *start
+    pub fn replace_data(&mut self, new_data: &Vec<u8>) {
+        unsafe {
```

Comment:
Check buffer size and border before touch it

Reply:
Changed the code to check if new_data can get in buffer